docs/2_2_1_fingerprinting.md

```diff
@@ -65,116 +65,244 @@ It is used to uniquely match a method by its characteristics.
-  parameters("Z")
-  ```
+    unrelatedMethod(string);
 
-- Method implementation:
+    // Filter 2, 3, 4 target instructions, and the instructions to modify.
+    if ("showBannerAds".equals(string)) {
+      showBannerAds();
+    }
 
-  ```kt
-  opcodes(Opcode.RETURN)
-  strings("pro")
-  ```
+    // Filter 5 and 6 target instructions.
+    return parameter2 != 1337;
+  }
 
-- Package and class name:
+  private void showBannerAds() {
+    // ...
+  }
 
-  ```kt
-  custom { (method, classDef) -> classDef == "Lcom/some/app/ads/AdsLoader;" }
-  ```
+  private void unrelatedMethod(String parameter) {
+    // ...
+  }
+}
+```
 
-With this information, the original code can be reconstructed:
+```asm
+# Method to fingerprint.
+.method public final obfuscatedMethod(Ljava/lang/String;ILObfuscatedClass;)Z
+    .registers 4
 
-```java
-package com.some.app.ads;
+    # Filter 1 target instruction.
+    sget-object v0, Lcom/some/app/ads/AdsLoader;->a:Ljava/util/Map;
 
-<accessFlags>
+    invoke-interface {v0, p1}, Ljava/util/Map;->get(Ljava/lang/Object;)Ljava/lang/Object;
 
-class AdsLoader {
-    public final boolean <methodName>(boolean <parameter>)
+    move-result-object p1
 
-    {
-        // ...
+    check-cast p1, Ljava/lang/String;
 
-        var userStatus = "pro";
+    invoke-direct {p0, p1}, Lcom/some/app/ads/AdsLoader;->unrelatedMethod(Ljava/lang/String;)V
 
-        // ...
+    # Filter 2 target instruction.
+    const-string v0, "showBannerAds"
 
-        return <returnValue >;
-    }
-}
+    # Filter 3 target instruction.
+    invoke-virtual {v0, p1}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+
+    # Filter 4 target instruction.
+    move-result p1
+
+    if-eqz p1, :cond_16
+
+    invoke-direct {p0}, Lcom/some/app/ads/AdsLoader;->showBannerAds()V
+
+    # Filter 5 target instruction.
+    :cond_16
+    const/16 p1, 0x539
+
+    # Filter 6 target instruction.
+    if-eq p2, p1, :cond_1c
+
+    const/4 p1, 0x1
+
+    goto :goto_1d
+
+    :cond_1c
+    const/4 p1, 0x0
+
+    :goto_1d
+    return p1
+.end method
 ```
 
-Using that fingerprint, this method can be matched uniquely from all other methods.
+  Notice the fingerprint filters do not declare every instruction in the target method, and between
+  each filter, zero or more other instructions can exist.  Instruction filters must be declared in
+  the same order as the instructions appear in the target method.
+
+  If the distance between each instruction declaration can be approximated, then the `maxAfter`
+  parameter can be used to restrict the instruction match to a maximum distance from the last
+  instruction.  A value of 0 for the first instruction filter means the filter must be the first
+  instruction of the target method.
+
+  If a single instruction varies slightly between different app targets but otherwise the fingerprint
+  is still the same, the `anyInstruction()` wrapper can be used to specify the different expected
+  instruction. Such as:
+  ```
+  anyInstruction(
```

Comment:
The DSL would be better if it was like this:


instructions {
 string() or opcode() // notice the "or" api
}

(or if we decide for non lambda (instructions(string() or opcode()))

Reply:
> Every time I see logical and arthemetic overloads in kotlin

The above can be expressed in one instruction { } block, but having operators like and or or, be applicable to the supertype "Check" makes it very dynamic, you can e.g. design a function that returns some condition, and then add your new condition "on top" . This is similar to typescripts typesystem:

MyType = Type & ({ newField } | { otherField })

Note how the type __checker__ can handle & and | arbitrarily.